### PR TITLE
Updated class to use WP filters instead.

### PR DESCRIPTION
**Pull request by ghost**
## Background

I am currently working in a Jade + PHP + Bootstrap framework and bumped into applying dropdown menus to Wordpress. After poking around and swallowing frustration I stumbled upon your class, but I couldn't quite wrap my head around it, so I began a 2 ½ day sleep deprivation struggle trying to make sense out of WP's filter / plugin framework and the `Walker` and `Walker_Nav_Menu` classes.
### Changes Summary
- Updated `start_lvl` to rely on the parent's implementation. This added one extra line but reduced the complexity of dealing with the internal indentation implementation details.
- `start_el` was removed as we can rely on WP's `nav_menu_css_class`,  `nav_menu_link_attributes` and `walker_nav_menu_start_el` filters to accomplish the same.
- `display_element` was slightly updated to be arguably more readable creating a `$this->has_children` property used throughout the filters via `$args->walker`.
- Added filters to customize the `$class_names`, `$atts` and final `$output` values. This reduced code duplication and size.
### Notes on Wordpress 4.0

WP 4.0 adds a new `protected $has_children;` property to `Walker_Nav_Menu`, effectively sparing us from hacking `display_element`.
### Rationale

After noticing your class was mostly duplicating `Walker_Nav_Menu` and adding the relevant parts to enable dropdown menus I edited out the redundancy and implemented my own solution via one function using filters and a trivial subclass of `Walker_Nav_Menu`. Actually, if there was a filter to customize the start of a level in the tree, we wouldn't even need to override `start_lvl` at all. This pull request is an adaption of my function to your class.
### tl;dr

It is more robust to rely on Wordpress' filter/hook framework than to subclass `Walker_Nav_Menu`. This avoids code duplication and arguably increases readability.

### CODE REVIEW

**Comment by SideQuestDIY** [comment]
This is an interesting approach. Thank you for your input. 

That said, I definitely cannot merge this in its current form as there are many features of the current walker that are missing from you adaptation.

We are working on a more robust version of the walker class and you approach may very well be the one we take. Feel free to take a look at the plan https://github.com/twittem/wp-bootstrap-navwalker/issues/84 and we would welcome your ideas. 

Any pull-requests made that alter the utilization and are not a simple bug tweek must first be committed to the dev branch for testing.

**Comment by ghost** [comment]
Thank you for getting back to me. I will definitely have a look at #84 and by the way, in case I am overlooking something, what are the missing features? As far as I can tell, the only thing I didn't add was:
- Support for any string case when using title_attr. (**Will commit later tomorrow.**)
- Making sure there were no indentation issues.

As for utilization, nothing has changed either. You still `require_once('wp-bootstrap-navwalker.php');` and pass an instance of the class when calling `wp_nav_menu`.

**Comment by ghost** [comment]
Now the class is not case sensitive to setting the title attribute. All my tests produce the same result as the original class and the usage remains the same.

But just in case:

> there are many features of the current walker that are missing from your adaptation.

Which ones?

**Comment by pattonwebz** [comment]
This PR errors out on me when I tested it. Getting this error:

`syntax error, unexpected '[' in /home/gamingn/public_html/wp-content/themes/breloaded/wp_bootstrap_navwalker.php on line 19`

This is the line causing the error:

`public function start_lvl( &$output, $depth = 0, $args = [] ) {`

Changing the `[]` to the `array()` function like this fixes the issue.

`public function start_lvl( &$output, $depth = 0, $args = array() ) {`

After making that small change and testing I can see no apparent front-end differences between this version of the walker and the original.

**Comment by pattonwebz** [comment]
I merged this into the `clean-dev` branch and added the fix I detailed above. 

@twittem if you wanted to test it just grab the copy from that branch and upload it to a site that used the current master version. Things should work with no modifications required. I really do think that this approach is the one we should take when finalizing the new version of the walker.

**Comment by ghost** [comment]
Cool, thanks. I see, so that's what @twittem meant. Maybe he didn't have enough time to look at it, but my version should work exactly like the original.

Are you using < PHP 5.4 perhaps?

**Comment by pattonwebz** [comment]
Yep you called it, seems the server that I tested on is running PHP 5.3 just... should probably update that to the latest version.

A quick look at the php changelog for 5.4 shows why it fails on 5.3 but will work on 5.4+ so essentially the change I made wasn't needed for up to date versions of PHP (but we can't rule out people on 5.3 because the server I tested on is built using webmin and that, by default, only installs php 5.3 for some reason).

**Comment by ghost** [comment]
Exactly. 5.3 goes back to 2009 and it's already unsupported [according to Wikipedia](http://en.wikipedia.org/wiki/PHP#Release_history). But `array()` is the way to go to support as many users as possible.

**Comment by SideQuestDIY** [comment]
To run WordPress your host just needs a couple of things:

PHP version 5.2.4 or greater
MySQL version 5.0 or greater

With this, a few things to keep in mind:
1. Whatever we build needs accessible on any platform that runs WordPress.
   Otherwise all we are doing it introducing new issues.
2. We need to strictly follow the WordPress coding standards. This was
   outlined so that everyone can understand. (Not saying this doesn't as I
   haven't been able to dig in that deep)

**Comment by pattonwebz** [comment]
Yeah it hit it's EOL back in August last year which is why I'm surprised to find that it was on my dev server at all. I only set that up that box about 2 weeks ago and installed webmin onto it so that the hosting stack on it was basically the same as the one I have on my production servers (I didn't notice till now that webmin only installs 5.3 by default - that's a little worrying if that's the case).

Since Ed pointed out that WP works with versions as far back as 5.2.4 we should do the same just for the sake of being compatible with as many hosts as possible.

**Comment by ghost** [comment]
:+1:

**Comment by SideQuestDIY** [comment]
Hey @bucaran,

If we can shore this up with the WP requirements I will gladly merge it will the existing project. Furthermore if you are interested in assisting with the project I would love to talk about that.

**Comment by ghost** [comment]
@twittem Do you mean these requirements?

> 1. Whatever we build needs accessible on any platform that runs WordPress.
>    Otherwise all we are doing it introducing new issues.
> 2. We need to strictly follow the WordPress coding standards. This was
>    outlined so that everyone can understand. (Not saying this doesn't as I
>    haven't been able to dig in that deep)
